### PR TITLE
[#19] [Backend] As a user, I can see a list of surveys on the home screen

### LINKED ISSUE #19: [Backend] As a user, I can see a list of surveys on the home screen
## Why

On the Home screen, the user should see the list of available surveys so that can choose one to begin the survey

## Acceptance Criteria

- Call `GET /api/v1/surveys` endpoint with the following query parameters:
  - `page[number]`
  - `page[size]`

- The list of surveys must be fetched when opening the application and **cached** onto the device.

_Note: Append l to the image URL to get the high-resolution image_
## Resources

[Postman - Nimble Survey API](https://documenter.getpostman.com/view/11835486/T1LJjTqj#2dd12ea2-3ecb-46d5-a48b-64256346e4e1)

### PR DESCRIPTION
**Pull request by markgravity**
https://github.com/markgravity/kmm-ic/issues/19

## What happened 👀

- Add `SurveyRepository`, `SurveySevice`, `CachedSurveyService` & `GetSurveysUseCase`
- Write UnitTest for `SurveyRepository` & `GetSurveysUseCase`

## Insight 📝

Please accept the unrelate file: [Package.resolved](https://github.com/markgravity/kmm-ic/pull/78#diff-0497307c97cedbd5c3599c03c8e39e0f7e5826c38887a3e3c1bea1702b7acd18), I discard the change on the local but the Xcode keeps update it to match with new package 🙏 

## Proof Of Work 📹

DEBUG LOG in iOS App
```
2023-04-26 17:18:52.282292+0700 KMMIC[43976:3499509] [AXRuntimeCommon] Unknown client: KMMIC
04-26 17:18:53.366 💚 DEBUG HttpClientCallLogger.closeResponseLog[async] - RESPONSE: 200 OK
METHOD: HttpMethod(value=GET)
FROM: https://survey-api.nimblehq.co/api/v1/surveys?page%5Bnumber%5D=1&page%5Bsize%5D=5
COMMON HEADERS
-> Alt-Svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400
-> Cache-Control: max-age=0, private, must-revalidate
-> Content-Encoding: br
-> Content-Type: application/json; charset=utf-8
-> Date: Wed, 26 Apr 2023 10:18:53 GMT
-> Etag: W/"e894f0cf1a9f2ea4689e8db96997a090"
-> Server: cloudflare
-> Vary: Accept, Accept-Encoding, Origin
-> Via: 1.1 vegur
-> cf-cache-status: DYNAMIC
-> cf-ray: 7bde158d7d36b46c-HKG
-> nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
-> referrer-policy: strict-origin-when-cross-origin
-> report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=hs%2BS7IoezTta8b%2FfPlOFAwuDz7v7zcd%2FuQGvLMI4d4b%2Bwor%2FkcuaCMO4Um%2F81NCJAX2ZHeJNxZ2xbZj5chtQMJdnei6pRG%2BAxGkQZPeYnQdnvj5lKTxjRJ9r%2BVguF1ZH%2B31Wu9vkcKF9Kr8HnyI4oJcoG1W1"}],"group":"cf-nel","max_age":604800}
-> x-content-type-options: nosniff
-> x-download-options: noopen
-> x-frame-options: SAMEORIGIN
-> x-permitted-cross-domain-policies: none
-> x-request-id: 268a6989-8d4b-4f5e-8c40-b4b916398137
-> x-runtime: 0.007363
-> x-xss-protection: 0
BODY Content-Type: application/json; charset=utf-8
BODY START
{"data":[{"id":"d5de6a8f8f5f1cfe51bc","type":"survey_simple","attributes":{"title":"Scarlett Bangkok","description":"We'd love ot hear from you!","thank_email_above_threshold":"\u003cspan style=\"font-family:arial,helvetica,sans-serif\"\u003e\u003cspan style=\"font-size:14px\"\u003eDear {name},\u003cbr /\u003e\u003cbr /\u003eThank you for visiting Scarlett Wine Bar \u0026amp; Restaurant at Pullman Bangkok Hotel G \u0026nbsp;and for taking the time to complete our guest feedback survey!\u003cbr /\u003e\u003cbr /\u003eYour feedback is very important to us and each survey is read individually by the management and owners shortly after it is sent. We discuss comments and suggestions at our daily meetings and use them to constantly improve our services.\u003cbr /\u003e\u003cbr /\u003eWe would very much appreciate it if you could take a few more
...
```

### CODE REVIEW

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/serializer/UrlSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+@Serializable(with = FullSizeUrlSerializer::class)
+class Url(val string: String)
+
+object FullSizeUrlSerializer : KSerializer<Url> {
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor("Url", PrimitiveKind.STRING)
+
+    override fun serialize(encoder: Encoder, value: Url) {
+        encoder.encodeString(value.string)
+    }
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/serializer/UrlSerializer.kt]: Optional: we can use the expression here.
```suggestion
    override fun serialize(encoder: Encoder, value: Url) =
        encoder.encodeString(value.string)
```

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/cached/survey/CachedSurveyService.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    override fun save(surveys: List<CachedSurvey>) {
+        realm.writeBlocking {
+            surveys.forEach {
+                copyToRealm(it, updatePolicy = UpdatePolicy.ALL)
+            }
+        }
+    }
+
+    override fun get(): List<CachedSurvey> {
+        return listOf()
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/cached/survey/CachedSurveyService.kt]: Why are we returning an empty list here? 🤔

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/model/Survey.kt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        surveyApiModel.coverImageUrl.string
+    )
+
+    internal constructor(cachedSurvey: CachedSurvey) : this(
+        cachedSurvey.id,
+        cachedSurvey.title,
+        cachedSurvey.description,
+        cachedSurvey.isActive,
+        cachedSurvey.coverImageUrl
+    )
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/domain/model/Survey.kt]: From a clean architecture perspective, it's recommended to keep the `domain` classes independent of the `data` and `presentation` ones. So it's better to move these factory methods to `SurveyApiModel` and `CachedSurvey` with the use of extension functions. Here's an example:
```
fun SurveyApiModel.toSurvey() = Survey(..)
```

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/SurveyRepositoryTest.kt [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+@ExperimentalCoroutinesApi
+class SurveyRepositoryTest {
+
+    @Mock
+    private val mockSurveyService = mock(classOf<SurveyService>())
+    @Mock
+    private val mockCachedSurveyService = mock(classOf<CachedSurveyService>())
+
+    private val mockThrowable = Throwable("mock")
+    private lateinit var repository: SurveyRepository
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/SurveyRepositoryTest.kt]: How about moving this property below the mock properties? Perhaps after a blank line?

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/SurveyRepositoryTest.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    private val mockSurveyApiModel = SurveyApiModel.dummy
+    private val mockCachedSurvey = CachedSurvey(mockSurveyApiModel)
+    private  val mockSurvey = Survey(mockSurveyApiModel)
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/data/repository/SurveyRepositoryTest.kt]: Extra whitespace. ⚠️ 
```suggestion
    private val mockSurvey = Survey(mockSurveyApiModel)
```

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetSurveysUseCaseTest.kt [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+import kotlinx.coroutines.test.runTest
+import org.koin.core.context.stopKoin
+import kotlin.test.AfterTest
+
+@ExperimentalCoroutinesApi
+class GetSurveysUseCaseTest {
+
+    @Mock
+    private val mockSurveyRepository = mock(classOf<SurveyRepository>())
+    private lateinit var useCase: GetSurveysUseCase
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetSurveysUseCaseTest.kt]: Same here. Keeping the SUT separate makes it more readable in terms of code organization. Also, how about renaming it to `getSurveysUseCase` to follow the naming convention?

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetSurveysUseCaseTest.kt [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+        useCase = GetSurveysUseCaseImpl()
+    }
+
+    @AfterTest
+    fun tearDown() {
+        stopKoin()
+    }
+
+    @Test
+    fun `when invoke is called - it returns the correctly surveys`() = runTest {
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetSurveysUseCaseTest.kt]:
```suggestion
    fun `when invoke is called - it returns the surveys correctly`() = runTest {
```

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetSurveysUseCaseTest.kt [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        given(mockSurveyRepository)
+            .function(mockSurveyRepository::getSurveys)
+            .whenInvokedWith(any())
+            .thenReturn(flowOf(listOf(mockSurvey)))
+
+        useCase(1, 1, false).collect {
+            it shouldBe listOf(mockSurvey)
+        }
```
Comment (AVI5HEK) [inline shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetSurveysUseCaseTest.kt]: It's just easier. 😆 
```suggestion
        useCase(1, 1, false).first() shouldBe listOf(mockSurvey)
```
Also, how about assigning `listOf(mockSurvey)` to a property and reusing that, e.g., `val expectedSurveys = listOf(mockSurvey)`

**Review by AVI5HEK — COMMENTED** [review]
Done with the first round.

File: shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetSurveysUseCaseTest.kt (hunk 6 above)
Comment (markgravity) [reply shared/src/commonTest/kotlin/co/nimblehq/mark/kmmic/domain/usecase/GetSurveysUseCaseTest.kt]: @AVI5HEK A simple name `useCase` here is enough to know this is `getSurveysUseCase` cause we are in `GetSurveysUseCaseTest`, IMO 🙏

**Review by suho — APPROVED** [review]

**Review by AVI5HEK — APPROVED** [review]
LGTM>

**Pull request merged**